compiler-rt/lib/fuzzer/FuzzerTracePC.cpp

```diff
@@ -629,20 +626,19 @@ void __sanitizer_weak_hook_strncmp(void *caller_pc, const char *s1,
   if (result == 0) return;  // No reason to mutate.
   size_t Len1 = fuzzer::InternalStrnlen(s1, n);
   size_t Len2 = fuzzer::InternalStrnlen(s2, n);
-  n = std::min(n, Len1);
-  n = std::min(n, Len2);
-  if (n <= 1) return;  // Not interesting.
-  fuzzer::TPC.AddValueForMemcmp(caller_pc, s1, s2, n, /*StopAtZero*/true);
+  // Not interesting.
+  if (Len1 <= 1 || Len2 <= 1)
+    return;
+  fuzzer::TPC.AddValueForMemcmp(caller_pc, s1, s2, Len1, Len2,
+                                /*StopAtZero*/ true);
 }
 
 ATTRIBUTE_INTERFACE ATTRIBUTE_NO_SANITIZE_MEMORY
 void __sanitizer_weak_hook_strcmp(void *caller_pc, const char *s1,
                                    const char *s2, int result) {
   if (!fuzzer::RunningUserCallback) return;
-  if (result == 0) return;  // No reason to mutate.
-  size_t N = fuzzer::InternalStrnlen2(s1, s2);
-  if (N <= 1) return;  // Not interesting.
-  fuzzer::TPC.AddValueForMemcmp(caller_pc, s1, s2, N, /*StopAtZero*/true);
+  return __sanitizer_weak_hook_strncmp(caller_pc, s1, s2,
+                                       fuzzer::Word::GetMaxSize(), result);
```

Comment:
Are we sure that it's valid to read up to `fuzzer::Word::GetMaxSize()` bytes from `s1` and `s2`? The caller probably expects no memory access beyond either pointer plus the length of the shorter string (or at most the length of the string).

Reply:
Note that the length is still passed through `InternalStrlen`, so we shouldn't read bytes beyond the end of either string.

But it's a good point that we may read bytes that a typical implementation of `strcmp` would not read. The C standard does not specify any particular behavior here, but of course that doesn't prevent people from depending on one. I don't know how to feel about this, but it may make upstreaming this difficult ;-)

Reply:
Indeed :(